src/common/src/hash/consistent_hash/vnode.rs

```diff
@@ -30,26 +30,44 @@ use crate::util::row_id::extract_vnode_id_from_row_id;
 pub struct VirtualNode(VirtualNodeInner);
 
 /// The internal representation of a virtual node id.
+///
+/// Note: not all bits of the inner representation are used.
 type VirtualNodeInner = u16;
-static_assertions::const_assert!(VirtualNodeInner::BITS >= VirtualNode::BITS as u32);
 
-impl From<Crc32HashCode> for VirtualNode {
-    fn from(hash_code: Crc32HashCode) -> Self {
+/// `vnode_count` must be provided to convert a hash code to a virtual node.
+///
+/// Use [`Crc32HashCodeToVnodeExt::to_vnode`] instead.
+impl !From<Crc32HashCode> for VirtualNode {}
+
+#[easy_ext::ext(Crc32HashCodeToVnodeExt)]
+impl Crc32HashCode {
+    fn to_vnode(self, vnode_count: usize) -> VirtualNode {
         // Take the least significant bits of the hash code.
         // TODO: should we use the most significant bits?
-        let inner = (hash_code.value() % Self::COUNT as u64) as VirtualNodeInner;
+        let inner = (self.value() % vnode_count as u64) as VirtualNodeInner;
         VirtualNode(inner)
     }
 }
 
 impl VirtualNode {
-    /// The number of bits used to represent a virtual node.
-    ///
-    /// Note: Not all bits of the inner representation are used. One should rely on this constant
-    /// to determine the count of virtual nodes.
-    pub const BITS: usize = 8;
     /// The total count of virtual nodes.
-    pub const COUNT: usize = 1 << Self::BITS;
+    // TODO(var-vnode): remove this and only keep `COUNT_FOR_TEST`
+    pub const COUNT: usize = 1 << 8;
+    /// The maximum value of the virtual node.
+    // TODO(var-vnode): remove this and only keep `MAX_FOR_TEST`
+    pub const MAX: VirtualNode = VirtualNode::from_index(Self::COUNT - 1);
```

Comment:
I think `COUNT` and `MAX` mean the same thing. Shall we merge them into one? Or do you plan to do it in the future PRs?

Reply:
Yes. They are just in different types. Will refactor in future PRs.